supercat/diffusion.py

```diff
@@ -127,36 +127,35 @@ def before_batch(self):
         else:
             alpha_bar_t = self.alpha_bar[t, None, None, None, None]
         alpha_bar_t = alpha_bar_t.to(self.dls.device)
-        
+
         # noisify the image
         xt =  torch.sqrt(alpha_bar_t) * hr + torch.sqrt(1-alpha_bar_t) * noise 
-        
-        # Stack input with low-resolution image (upscaled) and noise level
-        self.learn.xb = (torch.cat([xt, lr, alpha_bar_t.repeat(1,1,*hr.shape[2:])], dim=1),)
+
+        # Stack input with low-resolution image (upscaled) at channel dim,
+        # then pass the stacked image along with the noise level as tuple to the model
+        self.learn.xb = (torch.cat([xt, lr], dim=1), alpha_bar_t.view((batch_size, 1)))
         self.learn.yb = (noise,) # we are trying to predict the noise
 
 
 class DDPMSamplerCallback(DDPMCallback):
-    def before_batch(self):        
+    def before_batch(self):
         lr = self.xb[0]
-        
+        batch_size = lr.shape[0]
+
         # Generate a batch of random noise to start with
         xt = torch.randn_like(lr)
-        
-        outputs = [xt] 
+
+        outputs = [xt]
         for t in track(reversed(range(self.n_steps)), total=self.n_steps, description="Performing diffusion steps for batch:"):
             z = torch.randn(xt.shape, device=xt.device) if t > 0 else torch.zeros(xt.shape, device=xt.device)
             alpha_t = self.alpha[t] # get noise level at current timestep
             alpha_bar_t = self.alpha_bar[t]
             sigma_t = self.sigma[t]
-            model_input = torch.cat(
-                [xt, lr, alpha_bar_t.repeat(1,1,*lr.shape[2:]).to(xt.device)], 
-                dim=1,
-            )
-            predicted_noise = self.model(model_input)
-            
+
+            predicted_noise = self.model(torch.cat([xt, lr], dim=1), alpha_bar_t.view(batch_size, 1))
```

Comment:
Do we need to send the `alpha_bar_t` tensor to the same device as the input (i.e. the GPU if it is being used)?